eventuate-local-java-cdc-connector-mysql-binlog/src/test/java/io/eventuate/local/mysql/binlog/PerformanceTest.java

```diff
@@ -0,0 +1,367 @@
+      List<EntityTypeAndId> entityTypesAndIds = new ArrayList<>();
+
+      String entityType = generateId();
+      String entityId = generateId();
+
+      for (int i = 0; i < nEvents; i++) {
+        entityTypesAndIds.add(new EntityTypeAndId(entityType, entityId));
+      }
+
+      prepareTopics(entityTypesAndIds);
+      generateEvents(entityTypesAndIds);
+    });
+  }
+
+  @Test
+  public void testAllEventsDifferentTopicsSameId() throws Exception {
+    testPerformance(() -> {
+
+      List<EntityTypeAndId> entityTypesAndIds = new ArrayList<>();
+
+      String entityId = generateId();
+
+      for (int i = 0; i < nEvents; i++) {
+        entityTypesAndIds.add(new EntityTypeAndId(generateId(), entityId));
+      }
+
+      prepareTopics(entityTypesAndIds);
+      generateEvents(entityTypesAndIds);
+    });
+  }
+
+  @Test
+  public void test2TopicsSameId() throws Exception {
+    testPerformance(() -> {
+
+      List<EntityTypeAndId> entityTypesAndIds = new ArrayList<>();
+
+      String entityType = generateId();
+      String entityId = generateId();
+
+      for (int i = 0; i < nEvents; i++) {
+        entityTypesAndIds.add(new EntityTypeAndId(entityType + (i % 2), entityId));
+      }
+
+      prepareTopics(entityTypesAndIds);
+      generateEvents(entityTypesAndIds);
+    });
+  }
+
+  @Test
+  public void test4TopicsSameId() throws Exception {
+    testPerformance(() -> {
+
+      List<EntityTypeAndId> entityTypesAndIds = new ArrayList<>();
+
+      String entityType = generateId();
+      String entityId = generateId();
+
+      for (int i = 0; i < nEvents; i++) {
+        entityTypesAndIds.add(new EntityTypeAndId(entityType + (i % 4), entityId));
+      }
+
+      prepareTopics(entityTypesAndIds);
+      generateEvents(entityTypesAndIds);
+    });
+  }
+
+  @Test
+  public void test10TopicsSameId() throws Exception {
+    testPerformance(() -> {
+
+      List<EntityTypeAndId> entityTypesAndIds = new ArrayList<>();
+
+      String entityType = generateId();
+      String entityId = generateId();
+
+      for (int i = 0; i < nEvents; i++) {
+        entityTypesAndIds.add(new EntityTypeAndId(entityType + (i % 10), entityId));
+      }
+
+      prepareTopics(entityTypesAndIds);
+      generateEvents(entityTypesAndIds);
+    });
+  }
+
+  @Test
+  public void test100TopicsSameId() throws Exception {
+    testPerformance(() -> {
+      List<EntityTypeAndId> entityTypesAndIds = new ArrayList<>();
+
+      String entityType = generateId();
+      String entityId = generateId();
+
+      for (int i = 0; i < nEvents; i++) {
+        entityTypesAndIds.add(new EntityTypeAndId(entityType + (i % 100), entityId));
+      }
+
+      prepareTopics(entityTypesAndIds);
+      generateEvents(entityTypesAndIds);
+    });
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+  @Test
+  public void testAllEventsSameTopicDifferentIds() throws Exception {
+
```

Comment:
Each @Test seems be the same code over and over again with minor variations.

I think something like:

```
void runTest(int nEvents, Function<Integer, String> entityTypeGenerator, Supplier<String> idGenerator) {
    testPerformance(() -> {
       ...
   prepareTopics(entityTypesAndIds);
      generateEvents(entityTypesAndIds);
    });
  }
```   

along with various helpers

```
Function<Integer, String> makeFixedEntityTypeGenerator() {
   String entityType = generateId(); 
   return (i) -> entityType;
}

Function<Integer, String> makeNEntityTypeGenerator(int n) {
   String entityType = generateId(); 
   return (i) -> entityType + (i % n);
}

Supplier<String> makeUniqueIdSupplier() { 
   return () -> generateId(); 
}
```

etc

So that each test is

```
@Test
  public void test4TopicsDifferentIds() throws Exception {
    runTest(nEvents, makeNEntityTypeGenerator(4), makeUniqueIdSupplier()
  }
```

BTW why this: `throws Exception`

Reply:
Ok, I will fix,
Exception is not needed anymore.